### PR TITLE
#127, simplify friend selection description

### PR DESCRIPTION
**Pull request by cai-jack**
I added changes for #125 in Credits.tsx in this commit as well

### CODE REVIEW

**Comment by rbtying** [comment]
ahh, this one is a bit tricky: the skip counter is used in the backend to track the state of the friend-status, so nth isn't quite accurate. (test with picking the 2nd friend, and then watching what happens after the first card is played)

**Comment by cai-jack** [comment]
Got it - Skip counter is updated during play. will use it as the initial value only so that the ordinal number will not change during play. This is one of the more frequent complaints.

**Comment by rbtying** [comment]
I think the best way to do this would be to include another field in the backend for the ordinal number (i.e. modifying this: https://github.com/rbtying/shengji/blob/d126f0b8b5e8dc31ec966ec641f0edcecb3f0834/core/src/game_state.rs#L453)

**Comment by cai-jack** [comment]
@rbtying you are right. That is exactly what I just did. initial_skip was added to store the initial value which does not change during the course of play. Please check it out. Here is a screenshot.

Screen Shot 2020-06-07 at 12 26 30 PM

**Comment by cai-jack** [comment]
also, during exchange state, the "x played in previous hands" is not shown

File: frontend/src/Exchange.tsx [hunk 1]
```diff
@@ -124,10 +125,11 @@ class Exchange extends React.Component<IExchangeProps, IExchangeState> {
           <Trump trump={this.props.state.trump} />
           {this.props.state.game_mode !== 'Tractor' ? (
             <div>
-              <Friends gameMode={this.props.state.game_mode} />
+              <Friends gameMode={this.props.state.game_mode} showPlayed={false} />
               {this.state.friends.map((friend, idx) => {
                 const onChange = (x: IFriend) => {
                   const new_friends = [...this.state.friends];
+                  x.initial_skip = x.skip;
```
Comment (rbtying) [inline frontend/src/Exchange.tsx:132]: Let's only set `initial_skip` in the frontend (i.e. not `x.skip`), and in the backend set `skip = initial_skip`

https://github.com/rbtying/shengji/blob/d126f0b8b5e8dc31ec966ec641f0edcecb3f0834/core/src/game_state.rs#L1068

That way, the backend is protected against someone derping with the websocket (which people do...)

File: frontend/src/Friends.tsx [hunk 2]
```diff
@@ -17,22 +18,12 @@ const Friends = (props: Props) => {
...
-                {friend.skip} <InlineCard card={friend.card} /> can be played
-                before the next person to play <InlineCard card={friend.card} />{' '}
-                is a friend
-              </p>
-            );
-          }
+          return (
+            <p key={idx}>
+              The person to play the {nth(friend.initial_skip + 1)} <InlineCard card={friend.card} /> is a
+              friend. {(props.showPlayed) ? friend.initial_skip - friend.skip + ' played in previous hands' : ''}
```
Comment (rbtying) [inline frontend/src/Friends.tsx:24]: perhaps, "played in previous tricks"?

File: frontend/src/Friends.tsx [hunk 3]
```diff
@@ -41,4 +32,8 @@ const Friends = (props: Props) => {
   }
 };
 
+
+
+function nth(n: number) { return n + ["st", "nd", "rd"][(((n < 0 ? -n : n) + 90) % 100 - 10) % 10 - 1] || "th" }
```
Comment (rbtying) [inline frontend/src/Friends.tsx:37]: this is clever!

File: core/src/game_state.rs [hunk 4]
```diff
@@ -451,6 +451,7 @@ impl PropagatedState {
 pub struct Friend {
```
Comment (rbtying) [inline core/src/game_state.rs:451]: We can define another struct

```
pub struct FriendSelection {
  card: Card,
  initial_skip: usize,
}
```

and use that as an input into `set_friends`. In `set_friends` we can convert them into `Friend` structs.

File: frontend/src/Exchange.tsx (hunk 1 above)
Comment (cai-jack) [reply frontend/src/Exchange.tsx:132]: `friend` is a `&` reference, so the data it refers to cannot be written

File: frontend/src/Friends.tsx (hunk 3 above)
Comment (cai-jack) [reply frontend/src/Friends.tsx:37]: copied from stack Overflow :)

File: frontend/src/Exchange.tsx (hunk 1 above)
Comment (rbtying) [reply frontend/src/Exchange.tsx:132]: My suggestion would be something like:

```
pub struct FriendSelection {
  card: Card,
  initial_skip: usize,
}

pub fn set_friends(iter: impl IntoIterator<Item = FriendSelection>) -> ... {
  // ...
  let friends = iter.into_iter().map(|fs| Friend {
    card: fs.card,
    initial_skip: fs.initial_skip,
    skip: fs.initial_skip,
    player_id: None,
  });
}

```

File: frontend/src/Friends.tsx (hunk 2 above)
Comment (cai-jack) [reply frontend/src/Friends.tsx:24]: Sure. I will make it consistent with the current terms

File: core/src/game_state.rs (hunk 4 above)
Comment (cai-jack) [reply core/src/game_state.rs:451]: need to read more of your code, not clear how to convert from IFriend from wire (frontend)

Comment (rbtying) [reply core/src/game_state.rs:451]: if you derive `Serialize` and `Deserialize`, it should handle the serde for you.

You can look at `GameModeSettings` vs. `GameMode` for a similar pattern.

you'd need to update the corresponding enum in `interactive.rs` that describes an action to use `FriendSelection`, and the function signature for `set_friends`

**Comment by cai-jack** [comment]
@rbtying changes as you requested checked-in. Please verify. Note: I don't know how to use map() to set vec<FriendSelection> into vec<Friend>. I just used a simple iterator. Please let me know if you have a better idea.

File: frontend/src/Friends.tsx [hunk 5]
```diff
@@ -17,22 +18,12 @@ const Friends = (props: Props) => {
...
-                {friend.skip} <InlineCard card={friend.card} /> can be played
-                before the next person to play <InlineCard card={friend.card} />{' '}
-                is a friend
-              </p>
-            );
-          }
+          return (
+            <p key={idx}>
+              The person to play the {nth(friend.initial_skip + 1)} <InlineCard card={friend.card} /> is a
+              friend. {(props.showPlayed) ? friend.initial_skip - friend.skip + ' played in previous tricks' : ''}
```
Comment (rbtying) [inline frontend/src/Friends.tsx:24]: " were played" rather than "played"?

Comment (cai-jack) [reply frontend/src/Friends.tsx:24]: sentence vs phrase :) don't have to deal with singular/plural cases

**Pull request merged**